Refactor IAP Transport / Fix EASessions left open

kmicha19-ford opened the pull request:
Fixes #1799  (when used in conjunction with https://github.com/kmicha19-ford/sdl_ios/pull/2)
Fixes #1809
Fixes #1892
Fixes #1893

This PR is ready for review.

### Risk

This PR makes no  API changes.

### Testing Plan

- I have verified that I have not introduced new warnings in this PR (or explain why below)
- I have tested this PR against Core and verified behavior.

#### Unit Tests

Unit tests were updated for the refactored classes.

#### Core Tests

- Built the Obj-C example app and verified that all interactions work
   - iAP disconnect / reconnect
   - LockScreen on/off rapidly
- Built the Swift example app and verified that all interactions work
   - iAP disconnect / reconnect
   - LockScreen on/off rapidly

### Summary

* Refactors iAPSession code that uses the Apple External Accessory Frameworks to create, open, use, and close EASessions. 

### Changelog

##### Breaking Changes
* None

##### Enhancements

* Moves all managment of EASession creation, open, and closing into a single class, SDLIAPSession
* Eliminates the use of NSThread from SDLIAPDataSession
* SDLIAPSession now uses a DispatchQueue to provide the required thread for EASession
* Uses composition instead of inheritance, SDLIAPControlSession and SDLIAPDataSession now use SDLIAPSession instead of being subclasses of it
* Provides a short simple code path for closing EASessions
* Eliminates over two hundred lines of code

### Tasks Remaining:
- More unit tests

### CLA
- I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

NicoleYarroch commented on SmartDeviceLink/private/SDLIAPSession.m:
```diff
@@ -11,101 +11,285 @@
     }
-
+    
+    [stream close];
```
While testing, I have experienced random crashes at `[stream close]`. I suspect the issue is that `[stream close]` is being called on the main thread due to the accessory disconnected notification being sent on the main thread (in the `SDLIAPTransport` transport class). We have experienced crash in the [past](https://github.com/smartdevicelink/sdl_ios/pull/1592) and have found that we must open and close the I/O streams on the same thread. 

Since you switched from using a thread to dispatching to the `iapSessionQueue`, this means we have no guarantee that the I/O streams will be opened and closed on the same thread even if we change the code to dispatch closing the I/O threads to the `iapSessionQueue` (due to the fact that behind the scenes the blocks can be dispatched to different threads).

I was curious what the motivation was behind the switching from using a thread to to using the `iapSessionQueue`? Obviously, your changes are simpler and easier to read so I hope there is a solution that lets us keep the  `iapSessionQueue`.

kmicha19-ford replied:
iOS recommends that new designs not use NSThread and instead to use newer techologies such as GCD.  Using a DispatchQueue moves the responsility of creating and managing the thread to iOS. As you correctly mention, dispatch queue blocks can be dispatched to different threads. However self.eaSession always runs on the same thread, the thread that GCD used when we create self.eaSession here in SDLIAPSession

 (void) startSession {
dispatch_async(self.iapSessionQueue, ^{
        self.eaSession = [[EASession alloc] initWithAccessory:self.accessory forProtocol:self.protocolString];

Notice there are no other calls to dispatch_async(self.iapSessionQueue) anywhere in SDLIAPSession.  All further access is via self.eaSession, thus [stream close] is always on the same thread as was used for example in [stream open]

joeljfischer replied:
@kmicha19-ford I'm not quite following. Are you saying that `self.eaSession` is internally thread-safe and any calls to it will internally dispatch to the correct thread, and therefore it doesn't matter from where the `NSStream close` call comes from? Since those seem to be separate objects and classes, could you clarify the relation? I know an `EASession` uses the stream, but you seem to be closing the stream without accessing `self.eaSession` at this code line. As @NicoleYarroch noted, there seem to be crashes with this PR related to closing the stream, so it can't be entirely safe here.

makinney replied:
Sorry for delay with answer.  Work Mac went down yesterday. IT scheduled an appointment for Monday to take it in to run diagnostics. Meanwhile I am on a backup computer. 

EASession is not thread safe. As to other questions I need to get things setup on this backup Mac and so I can run this code and look at some things before answering these questions. Hopefully by Monday if not end of day today.

makinney replied:
@NicoleYarroch - Are you seeing random crashes at [stream close] with this PR? (Or something with existing code base)?

makinney replied:
@joeljfischer Which separate objects and classes are you referring to in your comment?

NicoleYarroch replied:
I was testing with the Swift example app using your branch. The crash occurred when I pulled the USB cord. I was testing with an iPhone 11 (iOS v.132.1). The crash only happened twice during ~30 disconnect tests. 

It was crashing at `[stream close]` both times. When I looked at the stack trace, I could see that the main thread was executing `[stream close]`. I traced it back to getting the accessory disconnect notification. 

I did notice while testing that sometimes `[stream close]` was triggered by the stream getting an `NSStreamEventEndEncountered` event and other times by the accessory disconnect notification. Sometimes both threads were trying to close the stream. It could be that the crash was related to both threads trying to close the stream at the same time? However I suspect that it is more likely that the the stream was being closed by a thread that does not own the stream (i.e. the main thread).

makinney replied:
I have a fix that ensures sessions are always closed on the sessions thread.  However there is an issue pushing the commit from my backup Mac. I am working on resolving this now.

kmicha19-ford replied:
@NicoleYarroch - I pushed a fix for close session.   Review appreciated.

NicoleYarroch reviewed (changes requested):
I mostly made some style formatting suggestions. Namely, private methods in the implementation files are prefixed by "sdl_". I tried to find most of them but I may have missed a few. 

I did experience a crash on disconnect that should be a pretty easy fix.

I haven't had the chance yet to thoroughly test secondary transports and switching between BT and USB transports but testing the multisession and control protocols worked well.

kmicha19-ford commented:
Thanks Nicole,

I have been without a work computer all week. Going into Dearborn today to get it fixed or replaced.  I will look at what you have here in detail by Monday if not by EOD today.

Mike
________________________________
From: Nicole Yarroch <notifications@github.com>
Sent: Wednesday, February 17, 2021 1:32 PM
To: smartdevicelink/sdl_ios <sdl_ios@noreply.github.com>
Cc: Kinney, Michael (K.A.) <kmicha19@ford.com>; Mention <mention@noreply.github.com>
Subject: Re: [smartdevicelink/sdl_ios] Refactor IAP Transport / Fix EASessions left open (#1910)


WARNING: This message originated outside of Ford Motor Company. Use caution when opening attachments, clicking links, or responding.


@NicoleYarroch requested changes on this pull request.

I mostly made some style formatting suggestions. Namely, private methods in the implementation files are prefixed by "sdl_". I tried to find most of them but I may have missed a few.

I did experience a crash on disconnect that should be a pretty easy fix.

I haven't had the chance yet to thoroughly test secondary transports and switching between BT and USB transports but testing the multisession and control protocols worked well.

________________________________

In SmartDeviceLink/private/SDLIAPControlSession.h<https://clicktime.symantec.com/3Gh5AeshoQfbTh4WeFQK5T47Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576811024>:

> @@ -32,7 +32,27 @@ NS_ASSUME_NONNULL_BEGIN

  *  @param delegate     The control session delegate

  *  @return             A SDLIAPControlSession object

  */

-- (instancetype)initWithAccessory:(nullable EAAccessory *)accessory delegate:(id<SDLIAPControlSessionDelegate>)delegate;

+- (instancetype)initWithAccessory:(nullable EAAccessory *)accessory delegate:(id<SDLIAPControlSessionDelegate>)delegate forProtocol:(NSString *)protocol;

+

+/**

+ *  Closes the SDLIAPSession used by the SDLIAPControlSession

+ */

+- (void) closeSession;

+

+/**

+ *  Returns whether the session has open I/O streams.

+ */

+@property (assign, nonatomic, readonly, getter=isSessionInProgress) BOOL sessionInProgress;

+

+/**

+ *  The accessory used to create and the EASession.



⬇️ Suggested change

- *  The accessory used to create and the EASession.

+ *  The accessory used to create the EASession.



________________________________

In SmartDeviceLink/private/SDLIAPControlSession.h<https://clicktime.symantec.com/3CCU6Q98i4pd6ixvNMe9nKE7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576811139>:

> @@ -32,7 +32,27 @@ NS_ASSUME_NONNULL_BEGIN

  *  @param delegate     The control session delegate

  *  @return             A SDLIAPControlSession object

  */

-- (instancetype)initWithAccessory:(nullable EAAccessory *)accessory delegate:(id<SDLIAPControlSessionDelegate>)delegate;

+- (instancetype)initWithAccessory:(nullable EAAccessory *)accessory delegate:(id<SDLIAPControlSessionDelegate>)delegate forProtocol:(NSString *)protocol;

+

+/**

+ *  Closes the SDLIAPSession used by the SDLIAPControlSession

+ */

+- (void) closeSession;



⬇️ Suggested change

-- (void) closeSession;

+- (void)closeSession;



________________________________

In SmartDeviceLink/private/SDLIAPSession.m<https://clicktime.symantec.com/379KLaGpXyn1Q2iFxF85wEv7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576812810>:

> +        return NO;

+        SDLLogD(@"EASession not in progress");



⬇️ Suggested change

-        return NO;

-        SDLLogD(@"EASession not in progress");

+        SDLLogD(@"EASession not in progress");

+        return NO;



________________________________

In SmartDeviceLink/private/SDLIAPSession.m<https://clicktime.symantec.com/38fWSK6eQVde8NXV4KKivvv7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576813674>:

>  }



 - (NSUInteger)connectionID {

     return self.eaSession.accessory.connectionID;

 }



+- (BOOL)hasSpaceAvailable {

+    return self.eaSession.outputStream.hasSpaceAvailable;

+}

+

+- (BOOL)isConnected {

+    return self.accessory.isConnected;

+}

+

+- (BOOL)isInputStream:(NSStream *)stream {

+    if (stream == self.eaSession.inputStream) {

+        return YES;

+    }

+    return NO;;



⬇️ Suggested change

-    return NO;;

+    return NO;



________________________________

In SmartDeviceLink/private/SDLIAPSession.m<https://clicktime.symantec.com/38p6aD1WUi34mZvK31GmNr57Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576813898>:

> +- (BOOL)isConnected {

+    return self.accessory.isConnected;

+}

+

+- (BOOL)isInputStream:(NSStream *)stream {

+    if (stream == self.eaSession.inputStream) {

+        return YES;

+    }

+    return NO;;

+}

+

+- (BOOL)isOutputStream:(NSStream *)stream {

+    if (stream == self.eaSession.outputStream) {

+        return YES;

+    }

+    return NO;;



⬇️ Suggested change

-    return NO;;

+    return NO;



________________________________

In SmartDeviceLink/private/SDLIAPSession.m<https://clicktime.symantec.com/3FnWdXrc7SwPfra5FPwGg8P7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576814887>:

> +- (BOOL)hasSpaceAvailable {

+    return self.eaSession.outputStream.hasSpaceAvailable;

+}

+

+- (BOOL)isConnected {

+    return self.accessory.isConnected;

+}

+

+- (BOOL)isInputStream:(NSStream *)stream {

+    if (stream == self.eaSession.inputStream) {

+        return YES;

+    }

+    return NO;;

+}

+

+- (BOOL)isOutputStream:(NSStream *)stream {



⬇️ Suggested change

-- (BOOL)isOutputStream:(NSStream *)stream {

+- (BOOL)sdl_isOutputStream:(NSStream *)stream {



________________________________

In SmartDeviceLink/private/SDLIAPSession.m<https://clicktime.symantec.com/3M6S3XWnXnTzAADDkEN6eqU7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576815779>:

>  }



 - (NSUInteger)connectionID {

     return self.eaSession.accessory.connectionID;

 }



+- (BOOL)hasSpaceAvailable {

+    return self.eaSession.outputStream.hasSpaceAvailable;

+}

+

+- (BOOL)isConnected {

+    return self.accessory.isConnected;

+}

+

+- (BOOL)isInputStream:(NSStream *)stream {



⬇️ Suggested change

-- (BOOL)isInputStream:(NSStream *)stream {

+- (BOOL)sdl_isInputStream:(NSStream *)stream {



________________________________

In SmartDeviceLink/private/SDLIAPSession.m<https://clicktime.symantec.com/3H11vB9J3xABKoJ8KkDDE7n7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576816368>:

>  }



 #pragma mark - Getters



-- (BOOL)isStopped {

-    return !self.isOutputStreamOpen && !self.isInputStreamOpen;

+- (BOOL)bothStreamsOpen {

+    if (self.inputStreamOpen && self.outputStreamOpen) {

+        return YES;

+    }

+    return NO;

+}

+

+- (BOOL)bothStreamsClosed {



⬇️ Suggested change

-- (BOOL)bothStreamsClosed {

+- (BOOL)sdl_bothStreamsClosed {



________________________________

In SmartDeviceLink/private/SDLIAPSession.h<https://clicktime.symantec.com/383gc1EiEw1E3gESPwfriT27Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576818008>:

>   */

-@property (nonatomic, assign) BOOL isInputStreamOpen;

+@property(readonly) BOOL bothStreamsOpen;



⬇️ Suggested change

-@property(readonly) BOOL bothStreamsOpen;

+@property(nonatomic, assign, readonly) BOOL bothStreamsOpen;



________________________________

In SmartDeviceLink/private/SDLIAPSession.h<https://clicktime.symantec.com/3WvQtRPESSh9TWbnCoUkjtP7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576819762>:

> +@end

+@interface SDLIAPSession : NSObject <NSStreamDelegate>



⬇️ Suggested change

-@end

-@interface SDLIAPSession : NSObject <NSStreamDelegate>

+@end

+

+/**

+ * Responsible for opening a connection with the accessory and transmitting data to and from the accessory. When the accessory disconnects, the connection is closed. Once the connection with the accessory is closed, the connection can not be reopened; instead, a new `SDLIAPSession` must be created.

+ */

+@interface SDLIAPSession : NSObject <NSStreamDelegate>



________________________________

In SmartDeviceLink/private/SDLIAPSession.h<https://clicktime.symantec.com/3JTyxLb3o6ifCN1qS2M4iQU7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576821244>:

>

 /**

- *  The unique protocol string used to create the session with the accessory.

+ *  Convenience initializer for setting an accessory and protocol string.



⬇️ Suggested change

- *  Convenience initializer for setting an accessory and protocol string.

+ *  Starts a session with the accessory.



________________________________

In SmartDeviceLink/private/SDLIAPSession.h<https://clicktime.symantec.com/3URDjxbyVfvkeFFhHq9x6V17Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576824594>:

>   */

-@property (assign, nonatomic, readonly, getter=isSessionInProgress) BOOL sessionInProgress;

+@property(readonly) BOOL hasSpaceAvailable;



⬇️ Suggested change

-@property(readonly) BOOL hasSpaceAvailable;

+@property(nonatomic, assign, readonly) BOOL hasSpaceAvailable;



________________________________

In SmartDeviceLink/private/SDLIAPSession.h<https://clicktime.symantec.com/38f9PmyfVcZHAZVMxscQ2jf7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576824675>:

>   */

-- (instancetype)initWithAccessory:(nullable EAAccessory *)accessory forProtocol:(NSString *)protocol;

+@property(readonly) BOOL isConnected;



⬇️ Suggested change

-@property(readonly) BOOL isConnected;

+@property(nonatomic, assign, readonly) BOOL isConnected;



________________________________

In SmartDeviceLink/private/SDLIAPSession.h<https://clicktime.symantec.com/3QCvPnm1j1sPai9ckwGv2JG7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576836491>:

> + *  Closes the EASession inputStream and the outputStream.

+ *  Sets bothStreamsOpen flag to false.

+ *  Stops SDLIAPSesssion operation by removing streams from the run loop

+ *  By design a SDLIAPSession instance cannot be reopened.



⬇️ Suggested change

- *  Closes the EASession inputStream and the outputStream.

- *  Sets bothStreamsOpen flag to false.

- *  Stops SDLIAPSesssion operation by removing streams from the run loop

- *  By design a SDLIAPSession instance cannot be reopened.

+ *  Closes the session's input and output streams. Once closed, the session can not be reopened.



________________________________

In SmartDeviceLink/private/SDLIAPSession.h<https://clicktime.symantec.com/3YbDHW2ZKGKuCr5jd5DtBaH7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576836946>:

>   */

-@property (nonatomic, assign) BOOL isOutputStreamOpen;

+- (void) closeSession;



⬇️ Suggested change

-- (void) closeSession;

+- (void)closeSession;



________________________________

In SmartDeviceLink/private/SDLIAPSession.h<https://clicktime.symantec.com/3EaiFDiPkUZmwABat3gMtJ37Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576837424>:

>   */

-- (void)stopStream:(NSStream *)stream;

-

-/**

- *  Cleans up a closed session

- */

-- (void)cleanupClosedSession;

-

+- (void)write:(NSMutableData *) data length: (NSUInteger) length  withCompletionHandler:(void (^)(NSInteger bytesWritten))completionHandler;



⬇️ Suggested change

-- (void)write:(NSMutableData *) data length: (NSUInteger) length  withCompletionHandler:(void (^)(NSInteger bytesWritten))completionHandler;

+- (void)write:(NSMutableData *)data length:(NSUInteger)length  withCompletionHandler:(void (^)(NSInteger bytesWritten))completionHandler;



________________________________

In SmartDeviceLink/private/SDLIAPDataSession.h<https://clicktime.symantec.com/3UeXfUSEBmSzvx2vfQRWGqo7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576840192>:

>

 - (instancetype)init NS_UNAVAILABLE;



+

+/**

+ *  Returns whether the session has open I/O streams.

+ */

+@property (assign, nonatomic, readonly, getter=isSessionInProgress) BOOL sessionInProgress;

+

+/**

+ *  The unique ID assigned to the session between the app and accessory. If no session exists the value will be 0.

+ */

+@property (assign, nonatomic, readonly) NSUInteger connectionID;

+

+/**

+ *  The accessory with which to open a session.

+ */

+@property (nullable, strong, nonatomic, readonly) EAAccessory *accessory;



⬇️ Suggested change

-@property (nullable, strong, nonatomic, readonly) EAAccessory *accessory;

+@property (nullable, strong, nonatomic, readonly) EAAccessory *accessory;



________________________________

In SmartDeviceLink/private/SDLIAPDataSession.h<https://clicktime.symantec.com/3L6sSGvtMkgVErRUCRNn1kh7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576840714>:

>

 - (instancetype)init NS_UNAVAILABLE;



+



⬇️ Suggested change

-



________________________________

In SmartDeviceLink/private/SDLIAPDataSession.m<https://clicktime.symantec.com/3TQiooNf4MRLYmTUCdpST9a7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576842231>:

> @@ -6,370 +6,144 @@

 //  Copyright © 2019 smartdevicelink. All rights reserved.

 //



+#import <UIKit/UIDevice.h>



⬇️ Suggested change

-#import <UIKit/UIDevice.h>



________________________________

In SmartDeviceLink/private/SDLIAPTransport.m<https://clicktime.symantec.com/3TXsK4rYqQsjWXAnZTsGQxE7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576866691>:

> @@ -57,6 +56,16 @@ - (instancetype)init {

     return self;

 }



+- (void) dealloc {



⬇️ Suggested change

-- (void) dealloc {

+- (void)dealloc {



________________________________

In SmartDeviceLink/private/SDLIAPTransport.m<https://clicktime.symantec.com/3UfsRVBiBKqskfyRvHMRLKA7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576870784>:

> +- (EAAccessory * _Nullable)accessory {

+    if (self.controlSession.accessory != nil) {

+        return self.controlSession.accessory;

+    }

+    if (self.dataSession.accessory != nil) {

+        return self.dataSession.accessory;

+    }

+    return nil;

+}



Appears to be unused.

________________________________

In SmartDeviceLink/private/SDLIAPTransport.m<https://clicktime.symantec.com/3E2TRufFiZWVmSVScgu3E1i7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576872086>:

> @@ -370,7 +386,7 @@ - (void)dataSessionDidReceiveData:(nonnull NSData *)data {

 /**

  *  Called when the data session should be retried.

  */

-- (void)dataSessionShouldRetry {

+- (void) dataSessionDidEnd{



⬇️ Suggested change

-- (void) dataSessionDidEnd{

+- (void)dataSessionDidEnd{



________________________________

In SmartDeviceLink/private/SDLIAPControlSession.m<https://clicktime.symantec.com/3NqFEpeEoG2zsDRu5Hao5Fx7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576891397>:

> @@ -22,206 +22,94 @@

 int const ProtocolIndexTimeoutSeconds = 10;



 @interface SDLIAPControlSession ()

-

 @property (nullable, strong, nonatomic) SDLTimer *protocolIndexTimer;

 @property (weak, nonatomic) id<SDLIAPControlSessionDelegate> delegate;



We should probably mark that delegate can be nil here

@property (nullable, weak, nonatomic) id<SDLIAPControlSessionDelegate> delegate;

________________________________

In SmartDeviceLink/private/SDLIAPControlSession.m<https://clicktime.symantec.com/3VzU5JjsCFuWcn5fHcnzEU17Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576892635>:

> -        __weak typeof(self) weakSelf = self;

-        [self sdl_startStreamsWithCompletionHandler:^(BOOL success) {

-            __strong typeof(weakSelf) strongSelf = weakSelf;

-            if (!success) {

-                SDLLogW(@"Control session failed to setup with accessory: %@. Attempting to create a new control session", strongSelf.accessory);

-                [strongSelf destroySessionWithCompletionHandler:^{

-                    __strong typeof(weakSelf) strongSelf = weakSelf;

-                    [strongSelf.delegate controlSessionShouldRetry];

-                }];

-            } else {

-                SDLLogD(@"Waiting for the protocol string from Core, setting timeout timer for %d seconds", ProtocolIndexTimeoutSeconds);

-                strongSelf.protocolIndexTimer = [strongSelf sdl_createControlSessionProtocolIndexStringDataTimeoutTimer];

-            }

-        }];

-    }

+- (void) closeSession {



⬇️ Suggested change

-- (void) closeSession {

+- (void)closeSession {



________________________________

In SmartDeviceLink/private/SDLIAPControlSession.m<https://clicktime.symantec.com/3MarjDvfjaGpQo42dB1dwu77Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576893340>:

> - *

- *  @param stream The stream that got the event code.

- */

-- (void)sdl_streamDidOpen:(NSStream *)stream {

-    if (stream == [self.eaSession outputStream]) {

-        SDLLogD(@"Control session output stream opened");

-        self.isOutputStreamOpen = YES;

-    } else if (stream == [self.eaSession inputStream]) {

-        SDLLogD(@"Control session input stream opened");

-        self.isInputStreamOpen = YES;

-    }

-

-    // When both streams are open, session initialization is complete. Let the delegate know.

-    if (self.isInputStreamOpen && self.isOutputStreamOpen) {

-        SDLLogV(@"Control session I/O streams opened for protocol: %@", self.protocolString);

+- (void) streamsDidOpen {



⬇️ Suggested change

-- (void) streamsDidOpen {

+- (void)streamsDidOpen {



________________________________

In SmartDeviceLink/private/SDLIAPControlSession.m<https://clicktime.symantec.com/3FdKtzPb8gBvfK8h3xAemkq7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576893617>:

>          [self sdl_startControlSessionProtocolIndexStringDataTimeoutTimer];

     }

 }



-/**

- *  Called when the session gets a `NSStreamEventEndEncountered` event code. The current session is closed and a new session is attempted.

- */

-- (void)sdl_streamDidEnd:(NSStream *)stream {

-    SDLLogD(@"Control stream ended");

-

-    // End events come in pairs, only perform this once per set.

-    [self.protocolIndexTimer cancel];

-

-    __weak typeof(self) weakSelf = self;

-    [self destroySessionWithCompletionHandler:^{

-        __strong typeof(weakSelf) strongSelf = weakSelf;

-        [strongSelf.delegate controlSessionShouldRetry];

-    }];

+- (void) streamsDidEnd {



⬇️ Suggested change

-- (void) streamsDidEnd {

+- (void)streamsDidEnd {



________________________________

In SmartDeviceLink/private/SDLIAPControlSession.m<https://clicktime.symantec.com/3hZXDkgSZBKLXJjdCLdBDJ7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576905628>:

>      // Read in the stream a single byte at a time

     uint8_t buf[1];

     NSInteger len = [inputStream read:buf maxLength:1];

     if (len <= 0) {

-        SDLLogV(@"No data in the control stream");

+        SDLLogD(@"No data in the control stream");



⬇️ Suggested change

-        SDLLogD(@"No data in the control stream");

+        SDLLogV(@"No data in the control stream");



________________________________

In SmartDeviceLink/private/SDLIAPDataSession.h<https://clicktime.symantec.com/3KkSR2211aiKx63R4fmdS2X7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576908960>:

> @@ -28,6 +42,11 @@ NS_ASSUME_NONNULL_BEGIN

  */

 - (instancetype)initWithAccessory:(nullable EAAccessory *)accessory delegate:(id<SDLIAPDataSessionDelegate>)delegate forProtocol:(NSString *)protocol;



+/**

+ *  Closes the SDLIAPSession used by the SDLIAPControlSession

+ */

+- (void) closeSession;



⬇️ Suggested change

-- (void) closeSession;

+- (void)closeSession;



________________________________

In SmartDeviceLink/private/SDLIAPDataSession.h<https://clicktime.symantec.com/3M3ZonagSND37oW11fLt5Uf7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576909124>:

> @@ -38,3 +57,4 @@ NS_ASSUME_NONNULL_BEGIN

 @end



 NS_ASSUME_NONNULL_END

+



⬇️ Suggested change

-



________________________________

In SmartDeviceLink/private/SDLIAPDataSession.m<https://clicktime.symantec.com/3M5mSYLA14mggDa33wLnQbY7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576909596>:

> -

-        // Attempt to cancel the ioStreamThread. Once the thread realizes it has been cancelled, it will cleanup the I/O streams. Make sure to wake up the run loop in case there is no current I/O event running on the ioThread.

-        [strongSelf.ioStreamThread cancel];

-        [strongSelf performSelector:@selector(sdl_doNothing) onThread:self.ioStreamThread withObject:nil waitUntilDone:NO];

-

-        // Block the thread until the semaphore has been released by the ioStreamThread (or a timeout has occured).

-        BOOL cancelledSuccessfully = [strongSelf sdl_isIOThreadCancelled];

-        if (!cancelledSuccessfully) {

-            SDLLogE(@"The I/O streams were not shut down successfully. We might not be able to create a new session with an accessory during the same app session. If this happens, only force quitting and restarting the app will allow new sessions.");

-        }

-

-        [strongSelf.sendDataQueue removeAllObjects];

-

-        disconnectCompletionHandler();

-    });

+- (void) closeSession {



⬇️ Suggested change

-- (void) closeSession {

+- (void)closeSession {



________________________________

In SmartDeviceLink/private/SDLIAPSession.m<https://clicktime.symantec.com/32aU9RRFeqWMynEh8mV2b5x7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576911133>:

> @@ -6,104 +6,297 @@

 #import "SDLLogMacros.h"

 #import "SDLMutableDataQueue.h"

 #import "SDLTimer.h"

+#import "SDLACVLLogging.h"



⬇️ Suggested change

-#import "SDLACVLLogging.h"



________________________________

In SmartDeviceLink/private/SDLIAPSession.m<https://clicktime.symantec.com/3CkUVW3F6ZMoATCW15diYgQ7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576911621>:

>  }



-- (void)startStream:(NSStream *)stream {

-    stream.delegate = self;

-    [stream scheduleInRunLoop:[NSRunLoop currentRunLoop] forMode:NSDefaultRunLoopMode];

-    [stream open];

+- (void) openStreams {



⬇️ Suggested change

-- (void) openStreams {

+- (void)openStreams {



________________________________

In SmartDeviceLink/private/SDLIAPSession.m<https://clicktime.symantec.com/3QEA1KmeyFNs7QGr4Ao4yx57Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576911765>:

>

-- (BOOL)createSession {

-    SDLLogD(@"Opening EASession with accessory: %@", self.accessory.name);

-    self.eaSession = [[EASession alloc] initWithAccessory:self.accessory forProtocol:self.protocolString];

-    return (self.eaSession != nil);

+- (void) startSession {



⬇️ Suggested change

-- (void) startSession {

+- (void)startSession {



________________________________

In SmartDeviceLink/private/SDLIAPSession.m<https://clicktime.symantec.com/3PWKG8SeDAgPQKSA8neyGbX7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576912612>:

>

-    NSUInteger status2 = stream.streamStatus;

-    if (status2 == NSStreamStatusClosed) {

-        if (stream == [self.eaSession inputStream]) {

-            SDLLogD(@"Input stream closed");

-                       self.isInputStreamOpen = NO;

-        } else if (stream == [self.eaSession outputStream]) {

-            SDLLogD(@"Output stream closed");

-                       self.isOutputStreamOpen = NO;

+- (void) closeSession {



⬇️ Suggested change

-- (void) closeSession {

+- (void)closeSession {



________________________________

In SmartDeviceLink/private/SDLIAPDataSession.h<https://clicktime.symantec.com/3GZ46uWjCkMUKyR7DUpK6nh7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576915162>:

> @@ -28,6 +42,11 @@ NS_ASSUME_NONNULL_BEGIN

  */

 - (instancetype)initWithAccessory:(nullable EAAccessory *)accessory delegate:(id<SDLIAPDataSessionDelegate>)delegate forProtocol:(NSString *)protocol;



+/**

+ *  Closes the SDLIAPSession used by the SDLIAPControlSession



⬇️ Suggested change

- *  Closes the SDLIAPSession used by the SDLIAPControlSession

+ *  Closes the SDLIAPSession used by the SDLIAPDataSession



________________________________

In SmartDeviceLink/private/SDLIAPDataSession.m<https://clicktime.symantec.com/3YK59WSBiCHKYJ1aPikwpGm7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576915833>:

>

+- (void) writeDataToSessionStream {



⬇️ Suggested change

-- (void) writeDataToSessionStream {

+- (void)sdl_writeDataToSessionStream {



________________________________

In SmartDeviceLink/private/SDLIAPDataSession.m<https://clicktime.symantec.com/3NUdhsbRiBZ5Ywxv3xb5fvW7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576916075>:

> -    NSInteger bytesWritten = [ostream write:remainder.bytes maxLength:bytesRemaining];

-

-    if (bytesWritten < 0) {

-        if (ostream.streamError != nil) {

-            // Once a stream has reported an error, it cannot be re-used for read or write operations. Shut down the stream and attempt to create a new session.

-            [self sdl_streamDidError:ostream];

-        } else {

-            // The write operation failed but there is no further information about the error. This can occur when disconnecting from an external accessory.

-            SDLLogE(@"Output stream write operation failed");

-        }

-    } else if (bytesWritten == bytesRemaining) {

-        // Remove the data from the queue

-        [self.sendDataQueue popBuffer];

+    if (remainder != nil) {

+        NSUInteger bytesRemaining = remainder.length;

+  //      SDLLogD(@"SDLIAPDataSession sending data for IAPSession %@.", self.iapSession);



⬇️ Suggested change

-  //      SDLLogD(@"SDLIAPDataSession sending data for IAPSession %@.", self.iapSession);



________________________________

In SmartDeviceLink/private/SDLIAPDataSession.m<https://clicktime.symantec.com/3GYKBKmSmgLfTSvz6iagqkB7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576916229>:

> -        } else {

-            // The write operation failed but there is no further information about the error. This can occur when disconnecting from an external accessory.

-            SDLLogE(@"Output stream write operation failed");

-        }

-    } else if (bytesWritten == bytesRemaining) {

-        // Remove the data from the queue

-        [self.sendDataQueue popBuffer];

+    if (remainder != nil) {

+        NSUInteger bytesRemaining = remainder.length;

+  //      SDLLogD(@"SDLIAPDataSession sending data for IAPSession %@.", self.iapSession);

+        [self.iapSession write:remainder length:bytesRemaining withCompletionHandler:^(NSInteger bytesWritten) {

+            if (bytesWritten == bytesRemaining) {

+                [self.sendDataQueue popBuffer];

+            } else {

+                // Cleave the sent bytes from the data, the remainder will sit at the head of the queue

+                // SDLLogD(@"SDLIAPDataSession writeDataToSessionStream bytes written %ld", (long)bytesWritten);



⬇️ Suggested change

-                // SDLLogD(@"SDLIAPDataSession writeDataToSessionStream bytes written %ld", (long)bytesWritten);



________________________________

In SmartDeviceLink/private/SDLIAPDataSession.m<https://clicktime.symantec.com/32Ag4f6QUe71tb6DRTebn8m7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576958723>:

> -            // The write operation failed but there is no further information about the error. This can occur when disconnecting from an external accessory.

-            SDLLogE(@"Output stream write operation failed");

-        }

-    } else if (bytesWritten == bytesRemaining) {

-        // Remove the data from the queue

-        [self.sendDataQueue popBuffer];

+    if (remainder != nil) {

+        NSUInteger bytesRemaining = remainder.length;

+  //      SDLLogD(@"SDLIAPDataSession sending data for IAPSession %@.", self.iapSession);

+        [self.iapSession write:remainder length:bytesRemaining withCompletionHandler:^(NSInteger bytesWritten) {

+            if (bytesWritten == bytesRemaining) {

+                [self.sendDataQueue popBuffer];

+            } else {

+                // Cleave the sent bytes from the data, the remainder will sit at the head of the queue

+                // SDLLogD(@"SDLIAPDataSession writeDataToSessionStream bytes written %ld", (long)bytesWritten);

+                [remainder replaceBytesInRange:NSMakeRange(0, (NSUInteger)bytesWritten) withBytes:NULL length:0];



I experienced a crash here several times. It looks like bytesWritten was -1 due to an error writing to the stream.

2021-02-16 11:11:35.670664-0500 SDL Example Swift[654:65963] ERROR - /BuildRoot/Library/Caches/com.apple.xbs/Sources/ExternalAccessory/ExternalAccessory-400.40.1/EAOutputStream.m:-[EAOutputStream write:maxLength:] - 306 failed to write 108 bytes (wrote -1) with error 32

2021-02-16 11:11:35.671106-0500 SDL Example Swift[654:67585] ERROR - /BuildRoot/Library/Caches/com.apple.xbs/Sources/ExternalAccessory/ExternalAccessory-400.40.1/EAOutputStream.m:-[EAOutputStream write:maxLength:] - 306 failed to write 108 bytes (wrote -1) with error 32

2021-02-16 11:11:35.983597-0500 SDL Example Swift[654:65963] *** Terminating app due to uncaught exception 'NSRangeException', reason: '*** -[NSConcreteMutableData replaceBytesInRange:withBytes:length:]: range {0, 18446744073709551615} exceeds data length 108'

*** First throw call stack:

(0x1aa44c96c 0x1aa165028 0x1aa77fdf8 0x1047b974c 0x104797d90 0x1047b9460 0x1047b92ec 0x1047745e0 0x104754088 0x104753b90 0x1047528fc 0x10466d748 0x10466c700 0x104aa327c 0x104ab2590 0x104ab1ddc 0x1046a7a40 0x10466c688 0x1046545d8 0x1aa816ffc 0x1aa3a499c 0x1aa3a49ec 0x1aa3a3ce4 0x1aa3a397c 0x1aa31c910 0x1aa3a32ac 0x1aa705b8c 0x1ae552fc4 0x1adf16500 0x1ae552d70 0x1ae561ce0 0x1adb4efec 0x1adb50164 0x1ae0e3b7c 0x1af642918 0x1af668ff8 0x1af64cef4 0x1af668f14 0x104aa327c 0x104aa69a4 0x1af68f254 0x1af68ef00 0x1af68f46c 0x1aa3c8108 0x1aa3c805c 0x1aa3c77c8 0x1aa3c2694 0x1aa3c1f40 0x1b4652534 0x1ae54d580 0x1041950ac 0x1aa240e18)

libc++abi.dylib: terminating with uncaught exception of type NSException

2021-02-16 11:11:35.983977-0500 SDL Example Swift[654:67585] *** Terminating app due to uncaught exception 'NSRangeException', reason: '*** -[NSConcreteMutableData replaceBytesInRange:withBytes:length:]: range {0, 18446744073709551615} exceeds data length 108'

*** First throw call stack:

(0x1aa44c96c 0x1aa165028 0x1aa77fdf8 0x1047b974c 0x104797d90 0x1047b9460 0x1047b99b0 0x10479a050 0x104799b7c 0x1c219cc64 0x1aa3d41b4 0x1aa3c8108 0x1aa3c805c 0x1aa3c7818 0x1aa3c2694 0x1aa3c1f40 0x1aa707340 0x104799a68 0x1047982c0 0x104aa1e1c 0x104aa327c 0x104aaa90c 0x104aab4fc 0x104ab64dc 0x1aa1576d0 0x1aa15d9e8)

*** Terminating app due to uncaught exception 'NSRangeException', reason: '*** -[NSConcreteMutableData replaceBytesInRange:withBytes:length:]: range {0, 18446744073709551615} exceeds data length 108'

terminating with uncaught exception of type NSException

CoreUI: deallocating _CUIInternalLinkRendition 205 /System/Library/CoreServices/CoreGlyphs.bundle/Assets.car



________________________________

In SmartDeviceLink/private/SDLIAPControlSession.m<https://clicktime.symantec.com/3DBdBTuL3s8wP2KYfF2Rn1e7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576960736>:

>  /**

  *  Called when the session gets a `NSStreamEventHasBytesAvailable` event code. A protocol string is created from the received data. Since a new session needs to be established with the protocol string, the current session is closed and a new session is created.

  */

-- (void)sdl_streamHasBytesAvailable:(NSInputStream *)inputStream {

-    SDLLogV(@"Control stream received data");

-

+- (void)streamHasBytesAvailable:(NSInputStream *)inputStream {

+    SDLLogD(@"SDLIAPControlSession EASession stream received data");



⬇️ Suggested change

-    SDLLogD(@"SDLIAPControlSession EASession stream received data");

+    SDLLogV(@"SDLIAPControlSession EASession stream received data");



________________________________

In SmartDeviceLink/private/SDLIAPDataSession.m<https://clicktime.symantec.com/34P4CwgCtmJt6cmVXydHER47Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576961588>:

> -    NSAssert(!NSThread.isMainThread, @"%@ should only be called on the IO thread", NSStringFromSelector(_cmd));

-

-    SDLLogE(@"Data session %s stream errored", stream == self.eaSession.inputStream ? "input" : "output");

-

-    // To prevent deadlocks the handler must return to the runloop and not block the thread

-    dispatch_async(dispatch_get_main_queue(), ^{

-        __weak typeof(self) weakSelf = self;

-        [self destroySessionWithCompletionHandler:^{

-            __strong typeof(weakSelf) strongSelf = weakSelf;

-            if (![strongSelf.protocolString isEqualToString:LegacyProtocolString]) {

-                if (strongSelf.delegate == nil) { return; }

-                [strongSelf.delegate dataSessionShouldRetry];

-            }

-        }];

-    });

+- (void) streamsDidEnd {



⬇️ Suggested change

-- (void) streamsDidEnd {

+- (void)streamsDidEnd {



________________________________

In SmartDeviceLink/private/SDLIAPDataSession.m<https://clicktime.symantec.com/3EEA933B39Lw5T4HAFPHKLe7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576961838>:

> -- (void)sdl_accessoryEventLoop {

-    @autoreleasepool {

-        NSAssert(self.eaSession != nil, @"Session must be assigned before calling");

-        if (!self.eaSession) {

-            return;

-        }

-

-        [self startStream:self.eaSession.inputStream];

-        [self startStream:self.eaSession.outputStream];

-

-        SDLLogD(@"Starting the accessory event loop on thread: %@", NSThread.currentThread.name);

-

-        while (self.ioStreamThread != nil && !self.ioStreamThread.cancelled) {

-            // Enqueued data will be written to and read from the streams in the runloop

-            [[NSRunLoop currentRunLoop] runMode:NSDefaultRunLoopMode beforeDate:[NSDate dateWithTimeIntervalSinceNow:0.25f]];

+- (void) streamDidError {



⬇️ Suggested change

-- (void) streamDidError {

+- (void)streamDidError {



________________________________

In SmartDeviceLink/private/SDLIAPSession.m<https://clicktime.symantec.com/3RdAocFgLGWs11vA9vtx2n47Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576963015>:

>

 NS_ASSUME_NONNULL_BEGIN



 @interface SDLIAPSession ()



-@property (nullable, strong, nonatomic, readwrite) EASession *eaSession;

+@property (nonatomic, weak) id<SDLIAPSessionDelegate> iAPSessionDelegate;



⬇️ Suggested change

-@property (nonatomic, weak) id<SDLIAPSessionDelegate> iAPSessionDelegate;

+@property (nullable, nonatomic, weak) id<SDLIAPSessionDelegate> iAPSessionDelegate;



________________________________

In SmartDeviceLink/private/SDLIAPSession.m<https://clicktime.symantec.com/3F6YqXYydNW2nhCesBRqc9y7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576973238>:

> -    if (status2 == NSStreamStatusClosed) {

-        if (stream == [self.eaSession inputStream]) {

-            SDLLogD(@"Input stream closed");

-                       self.isInputStreamOpen = NO;

-        } else if (stream == [self.eaSession outputStream]) {

-            SDLLogD(@"Output stream closed");

-                       self.isOutputStreamOpen = NO;

+- (void) closeSession {

+    bool waitUntilDone = NO;

+    if (NSThread.currentThread == self.sessionThread) {

+        waitUntilDone = YES;

+    }

+    [self performSelector:@selector(peformCloseSession) onThread:self.sessionThread withObject:nil waitUntilDone:waitUntilDone];

+}

+

+- (void) peformCloseSession {



⬇️ Suggested change

-- (void) peformCloseSession {

+- (void) sdl_peformCloseSession {



________________________________

In SmartDeviceLink/private/SDLIAPSession.m<https://clicktime.symantec.com/3MD2qXqjDdvkHV6MGZrDGPQ7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576973668>:

> +    [self performSelector:@selector(peformCloseSession) onThread:self.sessionThread withObject:nil waitUntilDone:waitUntilDone];

+}

+

+- (void) peformCloseSession {

+    SDLLogD(@"Closing EASession streams");

+    if (self.eaSession != nil) {

+        [self stopStreamRunLoop];

+        [self close: self.eaSession.inputStream];

+        [self close: self.eaSession.outputStream];

+        self.eaSession = nil;

+    } else {

+        SDLLogD(@"Failed to close streams because EASession is already nil");

+    }

+}

+

+- (void) startStreamRunLoop {



⬇️ Suggested change

-- (void) startStreamRunLoop {

+- (void)sdl_startStreamRunLoop {



________________________________

In SmartDeviceLink/private/SDLIAPSession.m<https://clicktime.symantec.com/3jcmEMKNAPChooFx91JUbj7Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23discussion_r576973820>:

>          }

     }

 }



-- (void)cleanupClosedSession {

-    if (self.eaSession == nil) {

-        SDLLogD(@"Attempting to close session with accessory: %@, but it is already closed", self.accessory.name);

-        return;

+- (void) stopStreamRunLoop {



⬇️ Suggested change

-- (void) stopStreamRunLoop {

+- (void)sdl_stopStreamRunLoop {



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://clicktime.symantec.com/3B6Hqi5nGWmnpaSZQw1wJB67Vc?u=https%3A%2F%2Fgithub.com%2Fsmartdevicelink%2Fsdl_ios%2Fpull%2F1910%23pullrequestreview-591208700>, or unsubscribe<https://clicktime.symantec.com/3Br9qDBNCXoXjRz5BaZzMu97Vc?u=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAIF2K33G322KTQHFY55UZP3S7QDVXANCNFSM4XDDDHJA>.

kmicha19-ford commented:
Just commited changes based on PR comments.  These include the following.  
prefixing all private methods with 'sdl_'
removed blank space from beginning of method names
removed extraneous semicolons
moved one public method up that was in the marked as private section
Note, not sure about when to you want to use SDLLogD, SDLLogV or even SDLLogE

NicoleYarroch commented on SmartDeviceLink/private/SDLIAPDataSession.m:
```diff
@@ -6,370 +6,144 @@
+            } else {
+                // Cleave the sent bytes from the data, the remainder will sit at the head of the queue
+                // SDLLogD(@"SDLIAPDataSession writeDataToSessionStream bytes written %ld", (long)bytesWritten);
+                [remainder replaceBytesInRange:NSMakeRange(0, (NSUInteger)bytesWritten) withBytes:NULL length:0];
```
@kmicha19-ford, were you able to fix this crash that I found? That was the major issue I wanted to get fix before I continued testing.

NicoleYarroch reviewed (changes requested):
I think a few things were missed from my last review. Can you tag me when you have a fix for the crash I noted in the last review? Thanks!

kmicha19-ford commented on SmartDeviceLink/private/SDLIAPDataSession.h:
```diff
@@ -7,18 +7,32 @@
+/**
+ *  The accessory with which to open a session.
+ */
+@property (nullable, strong, nonatomic, readonly) EAAccessory *accessory;
```
The suggest change looks the same as the original?

NicoleYarroch replied:
Sorry. Must have been a copy and paste error

kmicha19-ford commented on SmartDeviceLink/private/SDLIAPTransport.m:
```diff
@@ -224,9 +232,18 @@ - (void)disconnectWithCompletionHandler:(void (^)(void))disconnectCompletionHand
+    [self sdl_closeSessions:disconnectCompletionHandler];
 }
 
+- (EAAccessory * _Nullable)accessory {
+    if (self.controlSession.accessory != nil) {
+        return self.controlSession.accessory;
+    }
+    if (self.dataSession.accessory != nil) {
+        return self.dataSession.accessory;
+    }
+    return nil;
+}
```
Removed.  We use this in our SDL with OEM extensions to support so called marketplace apps.

kmicha19-ford commented:
I reviewed comments and made multiple updates for style requirements and also removed an used function.  Hopefully I have everything this time.

NicoleYarroch commented on SmartDeviceLink/private/SDLIAPDataSession.m:
```diff
@@ -7,369 +7,146 @@
-    }
-
-    SDLLogD(@"Closing EASession for accessory connection id: %tu, name: %@", self.connectionID, self.eaSession.accessory.name);
+#pragma mark Otherd
```
```suggestion
#pragma mark Getters
```

NicoleYarroch commented on SmartDeviceLink/private/SDLIAPTransport.m:
```diff
@@ -370,7 +376,7 @@ - (void)dataSessionDidReceiveData:(nonnull NSData *)data {
  *  Called when the data session should be retried.
  */
-- (void)dataSessionShouldRetry {
+- (void)dataSessionDidEnd{
```
```suggestion
- (void)dataSessionDidEnd {
```

NicoleYarroch commented on SmartDeviceLink/private/SDLIAPControlSession.m:
```diff
@@ -22,206 +22,94 @@
+    _iapSession = [[SDLIAPSession alloc] initWithAccessory:accessory forProtocol:protocol iAPSessionDelegate:self];
     _protocolIndexTimer = nil;
     _delegate = delegate;
+    SDLLogD(@"SDLIAPControlSession Wait for the protocol string from Core, setting timeout timer for %d seconds", ProtocolIndexTimeoutSeconds);
```
```suggestion
    SDLLogD(@"SDLIAPControlSession Waiting for the protocol string from Core, setting timeout timer for %d seconds", ProtocolIndexTimeoutSeconds);
```

NicoleYarroch reviewed (commented):
I found a few minor things to fix. A couple of comments from my previous reveiw were missed on the SDLIAPDataSession.m file.

NicoleYarroch reviewed (changes requested):
I found a few minor spacing issues - otherwise testing went well
